File: go/model_prices_test.go
package yir

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"testing"
)

func TestModelPriceInputs(t *testing.T) {
	raw, err := os.ReadFile("testdata/model-price-input-v1.json")
	if err != nil {
		t.Fatal(err)
	}
	var fixtures []struct {
		Name      string
		Operation string
		Request   GenerationRequest
		Expected  PriceInput
	}
	if err = json.Unmarshal(raw, &fixtures); err != nil {
		t.Fatal(err)
	}
	for _, f := range fixtures {
		t.Run(f.Name, func(t *testing.T) {
			got, err := BuildPriceInput(f.Operation, f.Request, "fixture-v1")
			if err != nil {
				t.Fatal(err)
			}
			a, _ := json.Marshal(got)
			b, _ := json.Marshal(f.Expected)
			if string(a) != string(b) {
				t.Fatalf("got %s; want %s", a, b)
			}
		})
	}
}

func TestGetModelPrices(t *testing.T) {
	table := PriceTable{Format: "yir-price-table-v1", Version: "v1", Purpose: "yir-cost", Unit: "USD", Scale: 1000000, Rows: []PriceRow{}}
	response := ModelPrices{ID: "klingai/kling-2.6", Object: "model_prices", Operation: "generate_video", InputMode: "text", ContractVersion: "c1", Coverage: "enumerated", Issues: []string{}, Prices: ModelPriceTables{Primary: table, Max: table, ExpiresAt: 2000000000, Unavailable: []ModelPriceUnavailable{}}}
	for _, expiry := range []int64{0, -1, maxPriceParameter + 1} {
		invalid := response
		invalid.Prices.ExpiresAt = expiry
		if err := invalid.Validate(response.ID, response.Operation, response.InputMode); err == nil {
			t.Fatalf("accepted invalid expiry %d", expiry)
		}
	}
	calls := 0
	query := ""
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls++
		if r.URL.RequestURI() != "/v1/models/klingai/kling-2.6?view=pricing&operation=generate_video&input_mode=text"+query || r.Header.Get("Authorization") != "Bearer test-key" {
			t.Error("wrong request")
		}
		json.NewEncoder(w).Encode(response)
	}))
	defer server.Close()
	client, err := NewClient("test-key", ClientOptions{BaseURL: server.URL})
	if err != nil {
		t.Fatal(err)
	}
	if _, err = client.GetModelPrices(context.Background(), response.ID, response.Operation, response.InputMode); err != nil {
		t.Fatal(err)
	}
	if _, err = client.GetModelPrices(context.Background(), "../private", response.Operation, response.InputMode); err == nil || calls != 1 {
		t.Fatal("invalid path reached server")
	}
	query = "&resolution=720p"
	response.Filter = &ModelPriceFilter{Resolution: "720p"}
	if _, err = client.GetModelPrices(context.Background(), response.ID, response.Operation, response.InputMode, ModelPriceFilter{Resolution: "720P"}); err != nil {
		t.Fatal(err)
	}
	response.Filter.Resolution = "1080p"
	if _, err = client.GetModelPrices(context.Background(), response.ID, response.Operation, response.InputMode, ModelPriceFilter{Resolution: "720p"}); err == nil {
		t.Fatal("accepted wrong filter")
	}
	count := 1
	before := calls
	if _, err = client.GetModelPrices(context.Background(), response.ID, response.Operation, response.InputMode, ModelPriceFilter{ReferenceCount: &count}); err == nil || calls != before {
		t.Fatal("invalid count reached server")
	}
	query = ""
	response.Filter = nil
	response.Prices.Primary.Purpose = "retail"
	if _, err = client.GetModelPrices(context.Background(), response.ID, response.Operation, response.InputMode); err == nil {
		t.Fatal("accepted retail data as cost")
	}
}

func TestModelPriceFilterScope(t *testing.T) {
	count := 2
	filter := ModelPriceFilter{Resolution: "2k", ReferenceCount: &count}
	conditions := map[string][]any{"input_mode": {"image"}, "resolution": {"2k"}}
	for _, role := range []string{"source_image", "first_frame", "last_frame", "reference_image", "reference_video", "reference_audio"} {
		conditions[role+"_count"] = []any{0}
	}
	conditions["reference_image_count"] = []any{2}
	row := PriceRow{ID: "scope", Model: "bytedance/seedream-5.0-pro", Operation: "generate_image", Conditions: conditions, Kind: "exact", Price: TablePrice{Type: "total", Amount: "1"}}
	table := PriceTable{Format: "yir-price-table-v1", Version: "v1", Purpose: "yir-cost", Unit: "USD", Scale: 1000000, Rows: []PriceRow{row}}
	response := ModelPrices{ID: row.Model, Object: "model_prices", Operation: row.Operation, InputMode: "image", ContractVersion: "c1", Coverage: "enumerated", Issues: []string{}, Filter: &filter, Prices: ModelPriceTables{Primary: table, Max: table, ExpiresAt: 2000000000, Unavailable: []ModelPriceUnavailable{}}}
	if err := response.Validate(row.Model, row.Operation, "image", filter); err != nil {
		t.Fatal(err)
	}
	if err := response.Validate(row.Model, row.Operation, "image"); err == nil {
		t.Fatal("filtered result accepted as complete model")
	}
	conditions["reference_image_count"] = []any{1}
	if err := response.Validate(row.Model, row.Operation, "image", filter); err == nil {
		t.Fatal("wrong row scope accepted")
	}
}

File: typescript/tests/model-prices.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import {readFileSync} from "node:fs";
import {buildPriceInput, calculatePrice, createYirClient} from "../dist/index.js";
import {validateModelPrices} from "../dist/shared/model-prices.js";
const fixtures = JSON.parse(readFileSync(new URL("../../go/testdata/model-price-input-v1.json", import.meta.url)));

test("filtered model prices preserve and validate the requested scope", async () => {
  const f=fixtures[1];
  const input=buildPriceInput(f.operation,f.request,"scope-v1");
  const filter={resolution:input.parameters.resolution,reference_count:2};
  const row={id:"scope",model:input.model,operation:f.operation,conditions:Object.fromEntries(Object.entries(input.parameters).map(([k,v])=>[k,[v]])),kind:"exact",price:{type:"total",amount:"1"}};
  const table={format:"yir-price-table-v1",version:"scope-v1",purpose:"yir-cost",unit:"USD",scale:1000000,rows:[row]};
  const response={id:input.model,object:"model_prices",operation:f.operation,input_mode:"image",contract_version:"c1",coverage:"enumerated",issues:[],filter,prices:{primary:table,max:table,expires_at:2000000000,unavailable:[]}};
  let calls=0;
  response.channel_parameters=[{provider:"apimart",channel_variant:"standard",operation:f.operation,input_mode:"image",parameter_rules:{quality:{behavior:"ignored",reason:"channel_parameter_unsupported",description:{zh:"不提供质量控制",en:"No quality control"}}}}];
  const client=createYirClient(async request=>{calls++;assert.ok(request.path.endsWith(`&resolution=${filter.resolution}&reference_count=2`));return structuredClone(response);});
  const got=await client.getModelPrices(input.model,f.operation,"image",{filter});
  assert.deepEqual(got.channel_parameters,response.channel_parameters);
  assert.equal(calculatePrice(got.prices.primary,input).amount,"1");
  assert.equal(calls,1);
  assert.throws(()=>validateModelPrices(response,input.model,f.operation,"image"),/response_invalid/);
  for(const bad of [ {...response,filter:{...filter,reference_count:1}}, {...response,filter:undefined} ]) assert.throws(()=>validateModelPrices(bad,input.model,f.operation,"image",filter),/response_invalid/);
  const wrong=structuredClone(response);
  wrong.prices.primary.rows[0].conditions.reference_image_count=[1];
  assert.throws(()=>validateModelPrices(wrong,input.model,f.operation,"image",filter),/response_invalid/);
  assert.throws(()=>client.getModelPrices(input.model,f.operation,"text",{filter}),/request_invalid/);
  assert.equal(calls,1);
});

for (const f of fixtures) test(`model price input: ${f.name}`, () => {
  const before = structuredClone(f.request);
  assert.deepEqual(buildPriceInput(f.operation, f.request, "fixture-v1"), f.expected);
  assert.deepEqual(f.request, before);
});

test("one model read supports repeated local prices and rejects unknown parameters", async () => {
  const f = fixtures[2];
  const conditions = Object.fromEntries(Object.entries(f.expected.parameters).map(([k,v])=>[k,[v]]));
  const table = {format:"yir-price-table-v1",version:"fixture-v1",purpose:"yir-cost",unit:"USD",scale:1000000,rows:[
    {id:"five",model:f.request.model,operation:f.operation,conditions,kind:"exact",price:{type:"total",amount:"184000"}},
    {id:"ten",model:f.request.model,operation:f.operation,conditions:{...conditions,duration:[10]},kind:"exact",price:{type:"total",amount:"368000"}},
  ]};
  const response = {id:f.request.model,object:"model_prices",operation:f.operation,input_mode:"text",contract_version:"test-v1",coverage:"enumerated",issues:[],prices:{primary:table,max:table,expires_at:2000000000,unavailable:[]}};
  for (const expires_at of [0,-1,Number.MAX_SAFE_INTEGER+1]) {
    assert.throws(()=>validateModelPrices({...response,prices:{...response.prices,expires_at}},f.request.model,f.operation,"text"),/response_invalid/);
  }
  let calls=0;
  const client=createYirClient(async request=> {
    calls++;
    assert.equal(request.method,"GET");
    assert.equal(request.path,"/v1/models/klingai/kling-2.6?view=pricing&operation=generate_video&input_mode=text");
    return structuredClone(response);
  });
  const model=await client.getModelPrices(f.request.model,f.operation,"text");
  for (const duration of [5,10,5,10]) {
    const request={...f.request,parameters:{...f.request.parameters,duration}};
    assert.equal(calculatePrice(model.prices.primary,buildPriceInput(f.operation,request,table.version)).amount,duration===5?"184000":"368000");
  }
  assert.equal(calls,1);
  const input=buildPriceInput(f.operation,f.request,table.version);
  assert.equal(calculatePrice(table,{...input,parameters:{...input.parameters,resolution:"1080p"}}).reason,"no_match");
  assert.throws(()=>buildPriceInput(f.operation,{...f.request,routing:{preference:"cost"}},table.version),/routing_override/);
  await assert.rejects(async()=>client.getModelPrices("../private",f.operation,"text"),/request_invalid/);
  assert.equal(calls,1);
  const bad=createYirClient(async()=>({...response,id:"another/model"}));
  await assert.rejects(bad.getModelPrices(f.request.model,f.operation,"text"),/response_invalid/);
});
